### PR TITLE
Caas storage detach

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

One liner - support detaching k8s storage instead of forcing it to be destroyed when removing a unit/application.

There's a large amount of iaas storage infrastructure for iaas clouds which was modified to also run with caas clouds. The essential stack is the same, but there's a number of differences to cater for:
- caas storage created/attached when the pod is provisioned, not separately
- the workers to monitor model scoped and unit scoped storage all run in the model manifold, not split across each application agent
- the workers just handle storage cleanup when things are deleted in the juju model

The work is broken up into several commits:

1. Allow caas storage to be detached instead of forcing destruction
Just a cleanup to revert the hard coding to destroy caas storage

2. Tweak state storage watchers and struct attachment tag to allow for watching unit storage
Just modify the regexp used with the existing state watchers to cater for unit attached storage as well as model attached storage.

3. Adapt the StorageProvisioner API facade
Modify the facade to support caas storage.

4. Adapt the StorageProvisioner worker
Modify the facade to support caas storage.

5. Add filesystem storage source to k8s broker
Implement some k8s FilesystemSource methods to delete volumes

6. Fix use of test kubernetes charms
We are now a lot stricter about loading fake k8s charms from the test repo.

## QA steps

deploy a k8s charm with storage
remove a unit and ensure the volume/filesystem are shown as detached
remove a unit is --destroy-storage ensure the volume/filesystem are deleted

### CODE REVIEW

**Comment by wallyworld** [comment]
!!build!!

**Comment by wallyworld** [comment]
!!build!!

File: state/volume.go [hunk 1]
```diff
@@ -941,15 +941,21 @@ func volumeAttachmentId(hostId, volumeName string) string {
...
 
 // ParseVolumeAttachmentId parses a string as a volume attachment ID,
-// returning the machine and volume components.
-func ParseVolumeAttachmentId(id string) (names.MachineTag, names.VolumeTag, error) {
+// returning the host and volume components.
+func ParseVolumeAttachmentId(id string) (names.Tag, names.VolumeTag, error) {
 	fields := strings.SplitN(id, ":", 2)
-	if len(fields) != 2 || !names.IsValidMachine(fields[0]) || !names.IsValidVolume(fields[1]) {
+	isValidHost := names.IsValidMachine(fields[0]) || names.IsValidUnit(fields[0])
+	if len(fields) != 2 || !isValidHost || !names.IsValidFilesystem(fields[1]) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/volume.go]: Right you are. Was a typo. Volume and filesystems coincidentally can have the same valid id.

File: state/watcher.go [hunk 2]
```diff
@@ -245,26 +247,33 @@ func (sb *storageBackend) watchModelMachinestorage(collection string) StringsWat
...
+// to the lifecycles of all filesystems scoped to units of the specified application.
+func (sb *storageBackend) WatchUnitFilesystems(app names.ApplicationTag) StringsWatcher {
+	return sb.watchHostStorage(app, filesystemsC)
 }
 
-func (sb *storageBackend) watchMachineStorage(m names.MachineTag, collection string) StringsWatcher {
+func (sb *storageBackend) watchHostStorage(host names.Tag, collection string) StringsWatcher {
 	mb := sb.mb
-	pattern := fmt.Sprintf("^%s/%s$", mb.docID(m.Id()), names.NumberSnippet)
+	pattern := fmt.Sprintf("^%s(/%s)?/%s$", mb.docID(host.Id()), names.NumberSnippet, names.NumberSnippet)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/watcher.go]: done

File: state/watcher.go [hunk 3]
```diff
@@ -305,27 +314,47 @@ func (sb *storageBackend) watchModelMachinestorageAttachments(collection string)
...
+	prefix := fmt.Sprintf("%s(/%s)?:%s(/%s)?/.*", host.Id(), names.NumberSnippet, host.Id(), names.NumberSnippet)
+	matchExp := regexp.MustCompile(prefix)
 	filter := func(id interface{}) bool {
 		k, err := mb.strictLocalID(id.(string))
 		if err != nil {
 			return false
 		}
-		return strings.HasPrefix(k, prefix)
+		matches := matchExp.FindStringSubmatch(k)
+		return len(matches) == 3 && matches[1] == matches[2]
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply state/watcher.go]: You might be off by 1?
See https://play.golang.org/p/tQYblc_tsRy

"mariadb/42:mariadb/42/666" ->
[mariadb/42:mariadb/42/666 /42 /42]
(match)

"mariadb/42:mariadb/41/666" ->
[mariadb/42:mariadb/41/666 /42 /41]
(no match)

File: apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go [hunk 4]
```diff
@@ -39,65 +39,94 @@ func (fw Watchers) WatchModelManagedFilesystems() state.StringsWatcher {
...
+		return w.loop()
+	})
+	return w
+}
+
 // WatchMachineManagedFilesystems returns a strings watcher that reports both
-// machine-scoped filesystems, and model-scoped, volume-backed filesystems
-// that are attached to the specified machine.
+// host-scoped filesystems, and model-scoped, volume-backed filesystems
+// that are attached to the specified host.
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go]: yeah it is

File: apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go [hunk 5]
```diff
@@ -39,65 +39,94 @@ func (fw Watchers) WatchModelManagedFilesystems() state.StringsWatcher {
 	})
 }
 
+func (fw Watchers) WatchUnitManagedFilesystems(app names.ApplicationTag) state.StringsWatcher {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go]: yup

File: apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go [hunk 6]
```diff
@@ -216,52 +249,82 @@ func (fw Watchers) WatchModelManagedFilesystemAttachments() state.StringsWatcher
...
 		defer watcher.Stop(w.modelFilesystemAttachments, &w.tomb)
 		defer watcher.Stop(w.modelVolumeAttachments, &w.tomb)
 		return w.loop()
 	})
 	return w
 }
 
-// machineFilesystemAttachmentsWatcher is a strings watcher that reports
-// lifechcle changes for attachments to both machine-scoped filesystems,
+func (fw Watchers) WatchUnitManagedFilesystemAttachments(app names.ApplicationTag) state.StringsWatcher {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/storageprovisioner/internal/filesystemwatcher/watchers.go]: no worries, should have done it in the first place

File: apiserver/facades/agent/storageprovisioner/storageprovisioner.go [hunk 7]
```diff
@@ -124,13 +131,16 @@ func NewStorageProvisionerAPIv3(
 					if canAccessStorageMachine(a.Host(), false) {
 						return true
 					}
+					return authorizer.AuthController()
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/storageprovisioner/storageprovisioner.go]: good pickup. this was left over from an older iteration. thanks

File: apiserver/facades/agent/storageprovisioner/storageprovisioner.go [hunk 8]
```diff
@@ -315,11 +349,18 @@ func (s *StorageProvisionerAPIv3) watchStorageEntities(
...
+		switch tag := tag.(type) {
+		case names.MachineTag:
 			w = watchMachineStorage(tag)
-		} else {
+		case names.ModelTag:
 			w = watchEnvironStorage()
+		case names.ApplicationTag:
+			w = watchApplicationStorage(tag)
 		}
+		if w == nil {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/storageprovisioner/storageprovisioner.go]: should have been default in the first place

File: worker/storageprovisioner/caasworker.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+			for i, appId := range apps {
+				appTags[i] = names.NewApplicationTag(appId)
+			}
+			appsLife, err := p.config.Life.Life(appTags)
+			if err != nil {
+				return errors.Annotate(err, "getting application life")
+			}
+			for i, appId := range apps {
+				appLifeResult := appsLife[i]
+				if appLifeResult.Error != nil && params.IsCodeNotFound(appLifeResult.Error) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply worker/storageprovisioner/caasworker.go]: any other error we don't specifically care about. just not found ie it's been deleted

File: worker/storageprovisioner/filesystem_events.go [hunk 10]
```diff
@@ -373,6 +375,10 @@ func processAliveFilesystemAttachments(
 		return errors.Trace(err)
 	}
 	for i, params := range params {
+		if params.Machine != nil && params.Machine.Kind() != names.MachineTagKind {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply worker/storageprovisioner/filesystem_events.go]: Right, that will be a big PR

File: caas/kubernetes/provider/storage_test.go [hunk 11]
```diff
@@ -101,3 +97,21 @@ func (s *storageSuite) TestScope(c *gc.C) {
 	p := s.k8sProvider(c, ctrl)
 	c.Assert(p.Scope(), gc.Equals, storage.ScopeEnviron)
 }
+
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply caas/kubernetes/provider/storage_test.go]: These aren't used and I'm not sure we will (I just added them for completeness). It depends if we want to allow importing into other models etc. I'll whip up a quick test.

**Comment by wallyworld** [comment]
$$merge$$

File: state/watcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (babbageclunk) [reply state/watcher.go]: Right - that's pulling out the unit number. But what I meant was: don't you also need to extract the application name part of the unit name? Or do you not need to check that the applications match here? (Maybe that's enforced at a higher-level?)

File: state/volume.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (babbageclunk) [reply state/volume.go]: Yeah, I figured they might be the same.

**Pull request merged**